QccPack-0.61-1/src/libQccPackWAV/utils/man1/wdrencode.1
.TH WDRENCODE 1 "QCCPACK" ""
.SH NAME
wdrencode, wdrdecode \- encode/decode image from/to PGM/PPM format using the original Wavelet Difference Reduction (WDR) algorithm
.SH SYNOPSIS
.B wdrencode 
.RB "[\|" \-w 
.IR wavelet "\|]"
.RB "[\|" \-b
.IR boundary "\|]" 
.RB "[\|" \-pw
.IR pcpfile "\|]" 
.RB "[\|" \-nl
.IR levels "\|]" 
.RB "[\|" \-m
.IR mask "\|]" 
.RB "[\|" \-vo "\|]"
.I rate imgfile bitstream

.B wdrdecode 
.RB "[\|" \-w 
.IR wavelet "\|]" 
.RB "[\|" \-b
.IR boundary "\|]" 
.RB "[\|" \-pw
.IR pcpfile "\|]" 
.RB "[\|" \-m
.IR mask "\|]"
.RB "[\|" \-r
.IR  rate "\|]"
.I bitstream imgfile
.SH DESCRIPTION
.TP
.BI \-w " wavelet"
String. Name of the wavelet file
.RB ( FBK
or 
.B LFT
for mat); default = "CohenDaubechiesFeauveau.9-7.lft".
.TP
.BI \-b " boundary"
String. Type of boundary extension; default = "symmetric".
.TP
.BI \-pw " pcpfile"
String. Name of file containing perceptual weights.
.TP
.BI \-nl " levels"
Integer. Number of levels of dyadic decomposition; default value is 5.
.TP
.BI \-m " mask"
String. Name of file containing the transperancy mask for shape-adaptive
coding.
.TP
.BI \-r " rate"
Float. Rate at which to stop decoding. Default is to decode until
end of input bitstream is reached.
.TP
.B \-vo
Print only the values of the rate to stdout (terse output); default is verbose output.
.SH DESCRIPTION
.B wdrencode
encodes an image, 
.IR imgfile , 
using the Wavelet Difference Reduction (WDR) algorithm by Tian and Wells.
.LP
The transform is either the usual discrete
wavelet transform (DWT) or a shape-adaptive
DWT (SA-DWT); see
.BR imgdwt (1)
for complete details.
Option
.B \-w
specifies the wavelet, 
option
.B \-b
specifies the boundary extension, and option
.B \-pw
indicates whether perceptual weighting is to be used.
The
.B \-m
option gives the image file to be used as a transparency mask
for the SA-DWT.
.I mask
must be a grayscale image
in which black in this image indicates transparency, white
indicates full opacity, and gray tones are degrees of
translucence
(see
.BR QccAlpha (3)).
If the
.B \-m
option is not given, the usual DWT is used.
.LP
The desired bitrate of the output
.I bitstream
is specified by
.IR rate ,
which is expressed in bits per pixel of the original image.
.B wdrencode
normally prints to stdout a small amount of information
concerning this target bitrate and the actual bitrate achieved.
Usually these rates are identical, but may differ slightly.
If the 
.B \-vo
(value-only) option is given, all printed output is suppressed except for
the actual bitrate.
.LP
.B wdrdecode
performs the corresponding decoding using the WDR algorithm, decoding
the
.I bitstream
made by
.B wdrencode
and producing
a reconstructed image,
.IR imgfile .
For sane results, 
.IR wavelet ,
.IR boundary ,
and
.I pcpfile
should be the same as they were for
.BR wdrencode ,
and
.I mask
should be the same if an SA-DWT was used as the transform.
If a decoding rate is specified with the
.B \-r
option, then
.B wdrdecode
stops decoding when
.I rate
bits per pixel from the input bitstream
have been decoded. Otherwise, decoding stops when the
end of the input bitstream is reached.
.LP
.B wdrencode
and
.B wdrdecode
call
.BR QccWAVwdrEncode (3) 
and 
.BR QccWAVwdrDecode (3), 
respectively, to perform WDR encoding and decoding. Refer to the man pages of these functions for more information.
.SH "SEE ALSO"
.BR QccWAVwdrEncode (3),
.BR QccWAVwdrDecode (3),
.BR imgdwt (1),
.BR imgdist (1),
.BR QccBitBuffer (3),
.BR QccENTArithmeticEncode (3),
.BR QccENTArithmeticDecode (3),
.BR QccWAVPerceptualWeights (3),
.BR QccWAVSubbandPyramid (3),
.BR QccPackWAV (3),
.BR QccPackIMG (3),
.BR QccPack (3)

J. Tian and R. O. Wells, Jr.,
"Embedded Image Coding Using Wavelet Difference Reduction", in
.IR "Wavelet Image and Video Compression" , 
P. N. Topiwala, Ed., pp. 289-302,
Kluwer Academic Publishers, Norwell, MA, 1998.

.SH AUTHOR
Written by Yufei Yuan <yuanyufei@hotmail.com>

Copyright (C) 1997-2016  James E. Fowler
.\"  The programs herein are free software; you can redistribute them and/or
.\"  modify them under the terms of the GNU General Public License
.\"  as published by the Free Software Foundation; either version 2
.\"  of the License, or (at your option) any later version.
.\"  
.\"  These programs are distributed in the hope that they will be useful,
.\"  but WITHOUT ANY WARRANTY; without even the implied warranty of
.\"  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
.\"  GNU General Public License for more details.
.\"  
.\"  You should have received a copy of the GNU General Public License
.\"  along with these programs; if not, write to the Free Software
.\"  Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.


QccPack-0.61-1/CHANGELOG
==============================================================================

                                  Q C C P A C K

                            Version 0.61, 01-apr-2016

                    Copyright (C) 1997-2016  James E. Fowler

------------------------------------------------------------------------------


CHANGES:

Version 0.61:
  01-apr-2016:  BETA release version 0.61
  30-mar-2016:  Updated Cygwin installation instructions
  20-jan-2016:  Updated copyright year range for 2016
  20-jan-2016:  Fixed zero-subband-size bug in tce3d.c

Version 0.60:
  25-nov-2015:  BETA release version 0.60
  08-aug-2015:  Added QccFileReadLine()
  07-aug-2015:  Added QccFileGetRealPath()
  14-jul-2015:  Added QccFileWriteMagicNumberVersion()
  
Version 0.59:
  18-jun-2015:  BETA release version 0.59
  15-jun-2015:  Added QccFileRewind()
  14-jun-2015:  Updated copyright year range for 2015
  14-jun-2015:  Made QccENTExponentialGolombEncodeSymbol() and
                  QccENTExponentialGolombDecodeSymbol() visible
  09-jan-2014:  Updated copyright year range for 2014
  20-jul-2013:  Added Data/Matlab to QCCPACK_DATA_DIRS in Imakefile
  20-jul-2013:  Updated copyright year range for 2013
  21-dec-2011:  Updated copyright year range for 2011
  21-dec-2011:  Minor fixes for warning-free compilation with gcc-4.6
  23-jun-2009:  Added better support for uniform quantizers to sqencode,
                  sqdecode
  11-mar-2009:  Added logo.sed

Version 0.58:
  13-feb-2009:  BETA release version 0.58
  05-feb-2009:  Added klttcedecode3d.1, klttcedecode3d_lossless.1, 
                  klttceencode3d.1, klttceencode3d_lossless.1, 
                  tcedecode3d_lossless.1, tceencode3d_lossless.1
  05-feb-2009:  Added QccWAVtce3DLosslessEncode.3,
                  QccWAVtce3DLosslessDecodeHeader.3, 
                  QccWAVtce3DLosslessDecode.3, 
                  QccWAVklttce3DEncode.3, 
                  QccWAVklttce3DDecodeHeader.3, 
                  QccWAVklttce3DDecode.3, 
                  QccWAVklttce3DLosslessEncode.3, 
                  QccWAVklttce3DLosslessDecodeHeader.3, 
                  QccWAVklttce3DLosslessDecode.3
  05-feb-2009:  Added klttce3d.c, klttceencode3d.c, klttcedecode3d.c
  04-feb-2009:  Changed QccHYPrkltTrain(), QccHYPrkltTransform(),
                  QccHYPrkltInverseTransform() to use integer volumes
  04-feb-2009:  Added klttce3d_lossless.c, klttceencode3d_lossless.c,
                  klttcedecode3d_lossless.c
  04-feb-2009:  Fixed zero temporal_num_scales problem in 3D packet transforms
  02-feb-2009:  Added tce3d_lossless.c,
                  tceencode3d_lossless, tcedecode3d_lossless
  29-jan-2009:  QccHYPrkltTrain.3, QccHYPrkltFactorization.3,
                  QccHYPrkltTransform.3, QccHYPrkltInverseTransform.3
  29-jan-2009:  Added QccHYPrklt.3
  29-jan-2009:  Added QccHYPrkltInitialize(), QccHYPrkltAlloc(),
                  QccHYPrkltFree(), QccHYPrkltTrain(),
                  QccHYPrkltFactorization()
  28-jan-2009:  Added QccHYPrkltTransform(), QccHYPrkltInverseTransform()
  28-jan-2009:  Added QccMatrixIdentity(), QccMatrixRowExchange(),
                  QccMatrixColExchange(), QccMatrixIdentity.3,
                  QccMatrixRowExchange.3, QccMatrixColExchange.3
  16-jan-2009:  Updated copyright year range for 2009

Version 0.57:
  30-jun-2008:  BETA release version 0.57
  09-may-2008:  Added QccTimeTic(), QccTimeToc(), QccTime.3
  06-may-2008:  Fixed spatial_num_levels/temporal_num_levels bug in
                  initialization in bisk3d.c, wdr3d.c
  06-may-2008:  Fixed zero number of levels in
                  QccWAVWaveletPacketDWT3D(),
                  QccWAVWaveletInversePacketDWT3D()
  26-apr-2008:  Added sam.c, icbsam.c, QccHYPImageCubeMeanSAM(),
                QccVectorAngle()
  17-apr-2008:  Added QccWAVtce3DEncode.3, tceencode3d.1
  16-apr-2008:  Added tce3d.c, tceencode3d.c, tcedecode3d.c
  16-apr-2008:  Added -bpv to rawtoicb.c, icbtoraw.c, rawdist3d.c
  16-apr-2008:  Added -bpp to rawtoicp.c, icptoraw.c, rawdist2d.c
  16-apr-2008:  Added bpv, bpp to QccHYPRaw(), QccHYPRawRead2D(),
                  QccHYPRawWrite2D(), QccHYPRawDist2D(), QccHYPRawRead3D(),
                  QccHYPRawWrite3D(), QccHYPRawDist3D(),

Version 0.56:
  14-mar-2008:  BETA release version 0.56
  14-mar-2008:  Added QccIMGImageComponentLBT.3,
                  QccIMGImageComponentInverseLBT.3
  13-mar-2008:  Added QccFastDCT, QccFastDCTInitialize, QccFastDCTCreate,
                  QccFastDCTFree, QccFastDCTForwardTransform1D,
                  QccFastDCTInverseTransform1D, QccFastDCTForwardTransform2D,
                  QccFastDCTInverseTransform2D
  13-mar-2008:  Added fast_dct.c
  13-mar-2008:  Added dcttceencode.1, dcttcedecode.1
  13-mar-2008:  Added QccWAVdcttceEncode.3, QccWAVdcttceDecodeHeader.3,
                  QccWAVdcttceDecode.3
  13-mar-2008:  Renamed image_lapped_filter.c to image_lbt.c
  24-feb-2008:  Fixed memory leaks in dcttce.c, image_lapped_filter.c
  17-jan-2008:  Added Haar.fbk
  09-jan-2008:  Added dcttce.c, dcttceencode.c, dcttcedecode.c
  09-jan-2008:  Added image_lapped_filter.c
  09-jan-2008:  Added QccMatrixInverse(), QccMatrixInverse.3
  04-jan-2008:  Updated substr
  04-jan-2008:  Updated copyright year range for 2008
  11-dec-2007:  Fixed minor max_coefficients bug in QccWAVtceEncodeDWT()
  24-oct-2007:  Changed spawn to be able to handle commands with arguments
  09-oct-2007:  Added QCCMATHEPS
  01-oct-2007:  Added QccMatrixIntMultiply(), QccMatrixIntMultiply.3
  01-oct-2007:  Added QccMatrixMultiply(), QccMatrixMultiply.3
  28-sep-2007:  Added QccMatrixNullspace(), QccMatrixNullspace.3
  28-sep-2007:  Added QccMatrixOrthogonalize(), QccMatrixOrthogonalize.3
  28-sep-2007:  Modified QccMatrixSVD() to work with matrics of any size

Version 0.55:
  29-aug-2007:  BETA release version 0.55
  03-jul-2007:  Added limits.h include
  17-may-2007:  Fixed num_symbols bug in QccENTHuffmanDesign()

QccPack-0.61-1/src/libQccPackWAV/utils/wdrencode.c
/*
 *
 * QccPack: Quantization, compression, and coding utilities
 * Copyright (C) 1997-2016  James E. Fowler
 * 
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; either version 2
 * of the License, or (at your option) any later version.
 * 
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 * 
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
 *
 */

/*
 *  This code was written by Yufei Yuan <yuanyufei@hotmail.com>
 */

#include "wdrencode.h"


#define USG_STRING "[-w %s:wavelet] [-b %s:boundary] [-pw %: %s:pcpfile] [-nl %d:num_levels] [-m %: %s:mask] [-vo %:] %f:rate %s:imgfile %s:bitstream"

QccWAVWavelet Wavelet;
QccString WaveletFilename = QCCWAVWAVELET_DEFAULT_WAVELET;
QccString Boundary = "symmetric";

QccWAVPerceptualWeights PerceptualWeights;
int UsePerceptualWeights = 0;

int NumLevels = 5;
int NumSubbands;

QccIMGImage InputImage;
int ImageNumRows, ImageNumCols;

QccIMGImage Mask;
int MaskSpecified = 0;
int MaskNumRows, MaskNumCols;

QccBitBuffer OutputBuffer;

int ValueOnly = 0;

float TargetRate;
float ActualRate;
int NumPixels;
int TargetBitCnt;


int main(int argc, char *argv[])
{
  int row, col;
  QccInit(argc, argv);
  
  QccWAVWaveletInitialize(&Wavelet);
  QccWAVPerceptualWeightsInitialize(&PerceptualWeights);
  QccIMGImageInitialize(&InputImage);
  QccIMGImageInitialize(&Mask);
  QccBitBufferInitialize(&OutputBuffer);
  
  if(QccParseParameters(argc, argv, USG_STRING,
                        WaveletFilename,
                        Boundary, 
                        &UsePerceptualWeights,
                        PerceptualWeights.filename,
                        &NumLevels,
                        &MaskSpecified,
                        Mask.filename,
                        &ValueOnly,
                        &TargetRate,
                        InputImage.filename,
                        OutputBuffer.filename))
    QccErrorExit();
  
  if (NumLevels < 0)
    {
      QccErrorAddMessage("%s: Number of levels of decomposition must be nonnegative",
                         argv[0]);
      QccErrorExit();
    }
  
  NumSubbands = QccWAVSubbandPyramidNumLevelsToNumSubbands(NumLevels);
  
  if (QccWAVWaveletCreate(&Wavelet, WaveletFilename, Boundary))
    {
      QccErrorAddMessage("%s: Error calling QccWAVWaveletCreate()", argv[0]);
      QccErrorExit();
    }
  
  if (UsePerceptualWeights)
    {
      if (QccWAVPerceptualWeightsRead(&PerceptualWeights))
        {
          QccErrorAddMessage("%s: Error calling QccWAVPerceptualWeightsRead()",
                             argv[0]);
          QccErrorExit();
        }
      if (PerceptualWeights.num_subbands != NumSubbands)
        {
          QccErrorAddMessage("%s: Number of subbands in %s does not match specified number of levels (%d)",
                             argv[0], PerceptualWeights.filename, NumLevels);
        }
    }
  
  if (QccIMGImageRead(&InputImage))
    {
      QccErrorAddMessage("%s: Error calling QccIMGImageRead()", argv[0]);
      QccErrorExit();
    }
  if (QccIMGImageColor(&InputImage))
    {
      QccErrorAddMessage("%s: Color images not supported", argv[0]);
      QccErrorExit();
    }
  if (QccIMGImageGetSize(&InputImage, &ImageNumRows, &ImageNumCols))
    {
      QccErrorAddMessage("%s: Error calling QccIMGImageGetSize()",
                         argv[0]);
      QccErrorExit();
    }
  
  if (MaskSpecified)
    {
      if (QccIMGImageRead(&Mask))
        {
          QccErrorAddMessage("%s: Error calling QccIMGImageRead()", argv[0]);
          QccErrorExit();
        }
      if (QccIMGImageColor(&Mask))
        {
          QccErrorAddMessage("%s: Mask must be grayscale", argv[0]);
          QccErrorExit();
        }
      if (QccIMGImageGetSize(&Mask, &MaskNumRows, &MaskNumCols))
        {
          QccErrorAddMessage("%s: Error calling QccIMGImageGetSize()",
                             argv[0]);
          QccErrorExit();
        }
      
      if ((MaskNumRows != ImageNumRows) || (MaskNumCols != ImageNumCols))
        {
          QccErrorAddMessage("%s: Mask must be same size as image", argv[0]);
          QccErrorExit();
        }
      
      NumPixels = 0;
      
      for (row = 0; row < MaskNumRows; row++)
        for (col = 0; col < MaskNumCols; col++)
          if (!QccAlphaTransparent(Mask.Y.image[row][col]))
            NumPixels++;
    }
  else
    NumPixels = ImageNumRows * ImageNumCols;
  
  OutputBuffer.type = QCCBITBUFFER_OUTPUT;
  if (QccBitBufferStart(&OutputBuffer))
    {
      QccErrorAddMessage("%s: Error calling QccBitBufferStart()", argv[0]);
      QccErrorExit();
    }
  
  TargetBitCnt = (int)(ceil((NumPixels * TargetRate)/8.0)) * 8;
  if (QccWAVwdrEncode(&(InputImage.Y),
                      (MaskSpecified ? &(Mask.Y) : NULL),
                      &OutputBuffer,
                      NumLevels,
                      &Wavelet,
                      ((UsePerceptualWeights) ? &PerceptualWeights : NULL), 
                      TargetBitCnt))
    {
      QccErrorAddMessage("%s: Error calling QccWAVwdrEncode()", argv[0]);
      QccErrorExit();
    }
  ActualRate = (double)OutputBuffer.bit_cnt / NumPixels;
  
  if (QccBitBufferEnd(&OutputBuffer))
    {
      QccErrorAddMessage("%s: Error calling QccBitBufferEnd()", argv[0]);
      QccErrorExit();
    }
  
  if (ValueOnly)
    printf("%f\n", ActualRate);
  else
    {
      printf("WDR coding of %s:\n", InputImage.filename);
      printf("  Target rate: %f bpp\n", TargetRate);
      printf("  Actual rate: %f bpp\n", ActualRate);
    }
  
  QccWAVWaveletFree(&Wavelet);
  QccWAVPerceptualWeightsFree(&PerceptualWeights);
  QccIMGImageFree(&InputImage);
  QccIMGImageFree(&Mask);

  QccExit;
}
